### PR TITLE
feat: Modify UI stream & team form 

### PR DESCRIPTION
**Pull request by shydefoo**
### Description
* This PR makes allowing the creation of new teams / streams on the UI configurable
* The default is still to allow new teams / streams to be added from the UI
* Modifies the validation of team and stream name to allow `_`, according to the allowed values described [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)

### CODE REVIEW

File: ui/packages/app/src/project_setting/form/ProjectForm.js [hunk 1]
```diff
@@ -13,7 +13,10 @@ import { addToast, EuiComboBoxSelect, useMlpApi } from "@caraml-dev/ui-lib";
 import { ProjectFormContext } from "./context";
 import { EmailTextArea } from "./EmailTextArea";
 import { Labels } from "./Labels";
-import { isDNS1123Label } from "../../validation/validation";
+import {
+  isDNS1123Label,
```
Comment (leonlnj) [inline ui/packages/app/src/project_setting/form/ProjectForm.js:17]: is this function still required at all?

Comment (shydefoo) [reply ui/packages/app/src/project_setting/form/ProjectForm.js:17]: seems like it's still being used for validating the custom labels added by users

**Comment by deadlycoconuts** [comment]
Sorry to ruin the party @shydefoo  but after #93 got merged, there are a couple of merge conflicts (but more importantly some refactoring to do) ><!

**Comment by shydefoo** [comment]
> Sorry to ruin the party @shydefoo but after #93 got merged, there are a couple of merge conflicts (but more importantly some refactoring to do) ><!

alright fixed

**Comment by shydefoo** [comment]
@deadlycoconuts @leonlnj poke

File: ui/packages/app/src/project_setting/form/Stream.js [hunk 2]
```diff
@@ -33,13 +33,22 @@ export const Stream = ({
...
-        onChange={onStreamChange}
-        onCreateOption={onStreamChange}
-        isDisabled={isDisabled}
-      />
+      {config.ALLOW_CUSTOM_STREAM ? (
+        <EuiComboBoxSelect
+          value={stream}
+          options={streamOptions}
+          onChange={onStreamChange}
+          onCreateOption={onStreamChange}
```
Comment (leonlnj) [inline ui/packages/app/src/project_setting/form/Stream.js:41]: is it possible to do the condition check here? like 

```
onCreateOption={config.ALLOW_CUSTOM_STREAM ? onStreamChange : undefined}
```

File: ui/packages/app/src/validation/validation.js [hunk 3]
```diff
@@ -23,3 +23,12 @@ export const validateEmail = email => {
   const expression = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;
   return expression.test(email);
 };
+
+// See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set for full details
+export const isValidK8sLabelValue = value => {
```
[the remark this answers is not in the document]
Comment (shydefoo) [reply ui/packages/app/src/validation/validation.js:28]: Ok apprently label keys also can have _ in them. If merging them is preferred i'll patch the isDNS1123Label function, probably name it something else since DNS1123 doesn't allow underscores

[the remark this answers is not in the document]
Comment (leonlnj) [reply ui/packages/app/src/validation/validation.js:28]: yea i thnik both sounded like its used for the same purpose and validation

File: ui/packages/app/src/project_setting/form/ProjectForm.js [hunk 4]
```diff
@@ -46,6 +46,12 @@ const ProjectForm = () => {
   };
 
   const [isValidStream, setIsValidStream] = useState(false);
+
+  useEffect(() => {
```
Comment (leonlnj) [inline ui/packages/app/src/project_setting/form/ProjectForm.js:50]: why is this needed while for labels and stream the `useEffect()` is not? Generally we should avoid `useEffect` unless necessary.

If the child component needs to know the validity, would something like this work? 

```
const isValidTeam = team => ...
 isValidTeam={isValidTeam(project.name)}
```

Comment (deadlycoconuts) [reply ui/packages/app/src/project_setting/form/ProjectForm.js:50]: Hmm if I understand correctly, I think it's just to check if the team's value is empty or not. In that case I think there's already this effect within the `Team` component: https://github.com/caraml-dev/mlp/blob/main/ui/packages/app/src/project_setting/form/Team.js#L34 so I wonder if we need this additional check here 🤔

File: ui/packages/app/src/project_setting/form/Stream.js [hunk 5]
```diff
@@ -37,8 +37,7 @@ export const Stream = ({
         value={stream}
         options={streamOptions}
         onChange={onStreamChange}
-        onCreateOption={onStreamChange}
-        isDisabled={isDisabled}
```
Comment (deadlycoconuts) [inline ui/packages/app/src/project_setting/form/Stream.js:41]: Can we not remove this `isDisabled` prop? 😅 Thanks 🙏

File: ui/packages/app/src/project_setting/form/Stream.js [hunk 6]
```diff
@@ -21,10 +21,10 @@ export const Stream = ({
   const [streamError, setStreamError] = useState("");
 
   const onStreamChange = stream => {
-    let isValid = isDNS1123Label(stream);
+    let isValid = isValidK8sLabelValue(stream);
     if (!isValid) {
       setStreamError(
-        "Stream name is invalid. It should contain only lowercase alphanumeric and dash (-), and must start and end with an alphanumeric character"
+        "Stream name is invalid. It should contain only lowercase alphanumeric and dash (-), or underscore (_) or period (.)"
```
Comment (deadlycoconuts) [inline ui/packages/app/src/project_setting/form/Stream.js:27]: Should we have `, and must start and end with an alphanumeric character"` added to the end of this string like for the error string in `Team.js`?

File: ui/packages/app/src/project_setting/form/Stream.js (hunk 5 above)
Comment (shydefoo) [reply ui/packages/app/src/project_setting/form/Stream.js:41]: fixed

File: ui/packages/app/src/project_setting/form/Stream.js (hunk 6 above)
Comment (shydefoo) [reply ui/packages/app/src/project_setting/form/Stream.js]: fixed

File: ui/packages/app/src/project_setting/form/ProjectForm.js (hunk 4 above)
Comment (shydefoo) [reply ui/packages/app/src/project_setting/form/ProjectForm.js]: yep @deadlycoconuts is right, this effect is already in `Team.js`, have removed it

**Review by leonlnj — APPROVED** [review]
thanks, lgtm!

**Comment by deadlycoconuts** [comment]
Thanks for the changes too! 🙏🏼🚀

**Review by deadlycoconuts — APPROVED** [review]
Thanks for the changes too! 🙏🏼🚀
